### PR TITLE
Svelte support

### PR DESCRIPTION
**Pull request by plumpNation**
Issue:  #2984 

## What I did
- Added support for Svelte.
- I added example kitchen sink for Svelte.
- Added *or tested* support for add ons:
  - centered
  - actions
  - background
  - knobs
  - links
  - notes
  - options
  - storyshots
  - storysource
  - viewport

## How to test
In root folder `yarn test`
Select `core` tests.

Is this testable with Jest or Chromatic screenshots? Yes, both
Does this need a new example in the kitchen sink apps? Yes
Does this need an update to the documentation? Yes

### CODE REVIEW

**Comment by igor-dv** [comment]
Oh, one more framework support, that's awesome!

File: examples/svelte-kitchen-sink/public/index.html
Comment (igor-dv) [inline examples/svelte-kitchen-sink/public/index.html]: React ?

File: ADDONS_SUPPORT.md
Comment (igor-dv) [inline ADDONS_SUPPORT.md]: please mark all the supported addons

File: addons/storyshots/storyshots-core/src/frameworks/svelte/loader.js
Comment (igor-dv) [inline addons/storyshots/storyshots-core/src/frameworks/svelte/loader.js]: I don't see any committed snapshots, does this addon really work?

File: ADDONS_SUPPORT.md
Comment (plumpNation) [reply ADDONS_SUPPORT.md]: I'm working my way through them, marking them off and committing as I go. I think I've marked off everything I've tested so far.

File: addons/storyshots/storyshots-core/src/frameworks/svelte/loader.js
Comment (plumpNation) [reply addons/storyshots/storyshots-core/src/frameworks/svelte/loader.js]: Not yet, I had to stop and sleep yesterday, got a bit late 😸

File: app/svelte/package.json
Comment (plumpNation) [inline app/svelte/package.json]: vue?

File: examples/vue-kitchen-sink/src/stories/addon-centered.stories.js
Comment (plumpNation) [inline examples/vue-kitchen-sink/src/stories/addon-centered.stories.js]: Not relevant to this review.

File: addons/centered/src/components/Centered.svelte
Comment (plumpNation) [reply addons/centered/src/components/Centered.svelte]: I can, but it will make the svelte template "worse". The style attribute doesn't work the same way as in react or vue, so I can't just pass the object in there.

I'll have a go, and see what you think.

Comment (plumpNation) [reply addons/centered/src/components/Centered.svelte]: I'm adding a small helper function to convert it to plain CSS strings before applying it inline in the component. Not very pretty but solves the problem. It doesn't make the component 'worse', so I guess there's that.

File: addons/storyshots/storyshots-core/src/frameworks/svelte/loader.js
Comment (plumpNation) [reply addons/storyshots/storyshots-core/src/frameworks/svelte/loader.js]: This should now be generating snapshots.

**Comment by ndelangen** [comment]
This is awesome @plumpNation !

File: README.md
Comment (igor-dv) [inline README.md]: Please place it after the HTML and add the `alpha` tag (for the chronological order)

File: README.md
Comment (igor-dv) [inline README.md]: Same here =)

File: examples/svelte-kitchen-sink/src/stories/__snapshots__/addon-actions.stories.storyshot
Comment (igor-dv) [inline examples/svelte-kitchen-sink/src/stories/__snapshots__/addon-actions.stories.storyshot]: I think this kind of snapshots is not really helping. You should probably implement the rendering function to really render the svelte components into html. 

I see you've added a transformer to the `jest.cofnig`, so probably you should make it something like in the `html/renderTree.js`

Comment (plumpNation) [reply examples/svelte-kitchen-sink/src/stories/__snapshots__/addon-actions.stories.storyshot]: I took the advice and implemented something similar to the html example. It seems to work.

File: .github/CODEOWNERS
Comment (Hypnosphi) [inline .github/CODEOWNERS]: Please mention yourself here

**Comment by Hypnosphi** [comment]
It changes from rounded to square as I click the button. This is probably not intended

https://teamcity.jetbrains.com/repository/download/OpenSourceProjects_Storybook_Svelte/1502394:id/svelte.zip%21/index.html?selectedKind=Button&selectedStory=rounded&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel

**Comment by igor-dv** [comment]
Please `yarn bootstrap` again to fix the lock file.

File: app/svelte/src/client/preview/globals.js
Comment (Hypnosphi) [inline app/svelte/src/client/preview/globals.js]: Shouldn't be needed here

File: app/svelte/src/client/preview/render.js
Comment (Hypnosphi) [inline app/svelte/src/client/preview/render.js]: Are those listeners automatically detached when component unmounts?

**Comment by Hypnosphi** [comment]
> This is probably not intended

Oh, looks like it actually is. Please make it more clear, e.g. using button label

File: examples/svelte-kitchen-sink/package.json
Comment (Hypnosphi) [inline examples/svelte-kitchen-sink/package.json]: Please update all the versions to alpha.10

**Comment by Hypnosphi** [comment]
Please apply changes from #3744

File: app/svelte/src/server/wrapInitialConfig.js
Comment (ekhaled) [inline app/svelte/src/server/wrapInitialConfig.js]: can we also support `.html` files here?

Comment (plumpNation) [reply app/svelte/src/server/wrapInitialConfig.js]: We can. I'm wondering if they will match other loaders elsewhere in the webpack config, haven't looked into these overrides enough yet.

File: app/svelte/src/client/preview/render.js
Comment (plumpNation) [reply app/svelte/src/client/preview/render.js]: I will look into it. I know some are, but they are internally managed within the components.

Comment (ekhaled) [reply app/svelte/src/client/preview/render.js]: Svelte removes those listeners... but the component has to be `destroy()`-ed.
Usually, svelte will take care of calling `destroy()` on components when they are removed from the DOM.
However, I'm not sure if you need to manually `destroy()` them when switching stories in this context

Comment (ekhaled) [reply app/svelte/src/client/preview/render.js]: Yeah, svelte will not remove any listeners added manually using `component.on()`.
They need to be removed manually.

```js
//add listener
const listener = component.on('someEvent', function(){});

//remove listener
listener.cancel();
```

File: addons/centered/src/components/Centered.svelte
Comment (ekhaled) [inline addons/centered/src/components/Centered.svelte]: Does not work in conjunction with `actions` as events are not forwarded from the inner component.

**Comment by c0bra** [comment]
Just wanted to drop a note to say that I would love to have this make it in. If there's any way to help I'm interested.

**Comment by igor-dv** [comment]
I think we are waiting for @plumpNation to finish the PR. Mostly to make the tests pass.

**Comment by ndelangen** [comment]
@c0bra I'm sure he's appreciate any help you could offer?

@plumpNation perhaps you could give @c0bra access to your fork, you your 2 can collaborate on it?

**Comment by plumpNation** [comment]
Sorry, I've been away. All help is appreciated :)

**Comment by plumpNation** [comment]
I've added @c0bra as a collaborator on my fork now. I have some more time to spend on this now. I'm hoping to mend the tests and supply generators, as well as address all comments.

Thanks for being patient.

File: app/svelte/src/client/preview/render.js
Comment (plumpNation) [reply app/svelte/src/client/preview/render.js]: For now I am not calling destroy on the component when changing stories, as the JS reference to it should be garbage collected, and the DOM nodes are removed by storybook.

I am adding something similar to what the Vue client/preview does, which is to keep a reference to the listeners and then cancel them when running a new render().

Thanks @ekhaled for the info.

File: app/svelte/src/client/preview/render.js
Comment (ekhaled) [inline app/svelte/src/client/preview/render.js]: I think wrap these in a `try/catch`.
listeners might be referencing some internal variable.. or DOM node, which got removed when the component was switched before this function could run

File: examples/official-storybook/stories/app-acceptance.stories.js
Comment (plumpNation) [reply examples/official-storybook/stories/app-acceptance.stories.js]: I'm sorry, I'm not sure what you mean.

I found a `built-storybooks.sh` where I have included this `svelte-kitchen-sink` also.

File: app/svelte/src/client/preview/render.js
Comment (plumpNation) [reply app/svelte/src/client/preview/render.js]: Calling `.cancel()` seems to just be spliced off the this._handlers array inside the component, if it still exists in that array. Would it still be a problem if the listener was referencing something that was non existent, since that code would not actually run?

Comment (ekhaled) [reply app/svelte/src/client/preview/render.js]: Actually, you are right. Because events supplied via `on` never actually correspond to actual DOM events.

And in this light, I don't think we need to cleanup these handlers at all.
Because the `_handlers` array will get de-referenced when the component is removed, and will be garbage collected.

However, we will need to call `destroy` on the components. So that they can remove their internal event handlers.
A component might have `<svelte:window on:resize='something()'>` components inside it. 
And these events will linger even after component was destroyed

File: addons/centered/src/components/Centered.svelte
Comment (plumpNation) [reply addons/centered/src/components/Centered.svelte]: This will be fixed in https://github.com/plumpNation/storybook/pull/1

File: app/svelte/src/client/preview/render.js
Comment (plumpNation) [reply app/svelte/src/client/preview/render.js]: Good hustle.

File: app/svelte/src/client/preview/render.js
Comment (plumpNation) [reply app/svelte/src/client/preview/render.js]: We decided on destroying the components rather than cancelling the listeners.

**Comment by ekhaled** [comment]
@plumpNation apart from https://github.com/storybooks/storybook/pull/3770#discussion_r200976975 is there any other outstanding issues that are blocking this?

File: ADDONS_SUPPORT.md
Comment (ekhaled) [inline ADDONS_SUPPORT.md]: `addon-a11y` works, but it shows an ugly error in the console.
```
Error: Expect axe._selectorData to be set up
    at generateSelector (axe.js:2810)
```
The error does not seem to affect functionality though

Comment (ndelangen) [reply ADDONS_SUPPORT.md]: Hmm that's unfortunate, I wonder if axe is actively maintained and we could open a ticket with the maintainers..

File: app/svelte/package.json
Comment (igor-dv) [inline app/svelte/package.json]: we are at 14 already 🤣

**Comment by plumpNation** [comment]
@ekhaled I don't think so. Missing features are not important at this stage; I should have introduced less in this PR.

File: app/svelte/src/server/wrapInitialConfig.js
Comment (ekhaled) [reply app/svelte/src/server/wrapInitialConfig.js]: I just tested.. and putting this:
```js
test: /\.(svelte|html)$/,
```
works without causing any issues with other loaders.

Maybe check it out from your side.

Comment (plumpNation) [reply app/svelte/src/server/wrapInitialConfig.js]: Thanks for checking that out. Will do.

**Comment by plumpNation** [comment]
@igor-dv do you have any tips for fixing the CI?

I can't figure out why teamcity is trying to spin up a node:latest linux docker on a Windows box.

https://teamcity.jetbrains.com/viewLog.html?tab=buildLog&buildTypeId=OpenSourceProjects_Storybook_Svelte&buildId=1538655

**Comment by tmeasday** [comment]
@ndelangen I think we just need to make sure the svelte example is built as part of this script: https://github.com/storybooks/storybook/blob/master/scripts/build-storybooks.sh

Then story can point at the built storybook!

**Comment by tmeasday** [comment]
Actually, not that is not right, that file is just for development.

I think the CI version of that happens inside Teamcity somewhere. Unfortunately the Teamcity API is pretty opaque to me, I'm not sure what you need to do that hasn't already been done, I suppose @Hypnosphi will have to say for sure.

**Comment by igor-dv** [comment]
Looks like TeamCity overcame its issues, by rerunning it few times..

**Comment by igor-dv** [comment]
Netlify works now: https://deploy-preview-3770--storybooks-svelte.netlify.com/

**Review by igor-dv — APPROVED** [review]
The last thing that is left is Chromatic, but I have no permissions in TC to play with settings...

**Comment by plumpNation** [comment]
@igor-dv thanks for the assist

**Review by ndelangen — APPROVED** [review]

**Comment by plumpNation** [comment]
From what I can see with the CodeFactor issues, these exist in all the other framework examples. If this is acceptable, is there anything else I need to do before this is merged into master?

**Comment by plumpNation** [comment]
Ah, the chromatic test.

**Comment by tmeasday** [comment]
Hey, to fix chromatic, you just need to change the file `examples/official-storybook/built-storybooks/svelte-kitchen-sink` to be a symlink to `../../svelte-kitchen-sink/storybook-static`

**Comment by ndelangen** [comment]
I created the symlink but this did not resolve the chromatic issue.

Also a unittest is failing:
```
 FAIL  examples/svelte-kitchen-sink/svelteshots.test.js
  ● Test suite failed to run

    storyshots is intended only to be used with storybook

      2 | import initStoryshots, { multiSnapshotWithOptions } from '@storybook/addon-storyshots';
      3 |
    > 4 | initStoryshots({
        | ^
      5 |   framework: 'svelte',
      6 |   configPath: path.join(__dirname, '.storybook'),
      7 |   integrityOptions: { cwd: path.join(__dirname, 'src', 'stories') },

      at loadFramework (examples/svelte-kitchen-sink/node_modules/@storybook/addon-storyshots/dist/frameworks/frameworkLoader.js:42:11)
      at testStorySnapshots (examples/svelte-kitchen-sink/node_modules/@storybook/addon-storyshots/dist/api/index.js:54:54)
      at Object.<anonymous> (examples/svelte-kitchen-sink/svelteshots.test.js:4:1)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        2.13s
Ran all test suites matching /svelteshots.test.js/i.

Watch Usage: Press w to show more.
```

**Comment by igor-dv** [comment]
It is because a wrong alpha versions after the merge. Everything should be 16

**Comment by igor-dv** [comment]
Ok, I think we've tried everything to solve chromatic. I've even tried to exorcise my laptop... 
Chromatic people, can you please teach me how to do this?

**Comment by tmeasday** [comment]
Just as a note -- I could get the Acceptance story running properly locally when i removed the file at `examples/official-storybook/built-storybooks/svelte-kitchen-sink` and recreated the symlink. However git was confused. 

I suspect the solution is to remove the file that currently exists, commit, recreate the symlink, then commit again. I can do it if you like -- not sure where to commit to?

**Comment by igor-dv** [comment]
This file was my attempt to create a symlink on windows 😓 
I've deleted it now.

**Comment by tmeasday** [comment]
Sorry I didn't realise I could push to the fork. Did that, hopefully fixed now.

**Comment by tmeasday** [comment]
Ugh, Ok hopefully did it properly this time.

**Comment by igor-dv** [comment]
hellel

**Comment by plumpNation** [comment]
Yes! Thank you so much guys. Really appreciate you not only accepting it, but helping out to push it out too.

**Pull request merged**
